File: ValidPalindrome.java
import java.util.ArrayList;
import java.util.List;

public class ValidPalindrome {
    public static boolean isPalindrome(String s) {
        s = s.replaceAll("[^A-Za-z0-9]", "").toLowerCase();
        int start = 0;
        int end = s.length() - 1;
        while (start < end) {
            if (s.charAt(start) != s.charAt(end)) {
                return false;
            }
            start++;
            end--;
        }
        return true;
    }

    public static void main(String[] args) {
        testIsPalindrome();
    }

    public static void testIsPalindrome() {
        List<TestCase> testCases = new ArrayList<>();
        testCases.add(new TestCase("A man, a plan, a canal, Panama", true));
        testCases.add(new TestCase("race a car", false));
        testCases.add(new TestCase("Madam, in Eden, I'm Adam", true));

        for (int i = 0; i < testCases.size(); i++) {
            TestCase testCase = testCases.get(i);
            boolean result = isPalindrome(testCase.input);
            if (result == testCase.expected) {
                System.out.println("Test Case " + (i + 1) + ": Passed");
            } else {
                System.out.println("Test Case " + (i + 1) + ": Failed");
                System.out.println("  Expected: " + testCase.expected);
                System.out.println("  Received: " + result);
            }
        }
    }

    static class TestCase {
        String input;
        boolean expected;

        TestCase(String input, boolean expected) {
            this.input = input;
            this.expected = expected;
        }
    }
}

File: valid_palindrome.py
def is_palindrome(s):
    s = ''.join(c for c in s if c.isalnum()).lower()
    start = 0
    end = len(s) - 1
    while start < end:
        if s[start] != s[end]:
            return False
        start += 1
        end -= 1
    return True

def test_is_palindrome():
    test_cases = [
        {"input": "A man, a plan, a canal, Panama", "expected": True},
        {"input": "race a car", "expected": False},
        {"input": "Madam, in Eden, I'm Adam", "expected": True}
    ]

    for index, test_case in enumerate(test_cases):
        result = is_palindrome(test_case["input"])
        if result == test_case["expected"]:
            print(f"Test Case {index + 1}: Passed")
        else:
            print(f"Test Case {index + 1}: Failed")
            print(f"  Expected: {test_case['expected']}")
            print(f"  Received: {result}")

test_is_palindrome()